Edit and Remove Enrollments (Backend)

zPereiraa opened the pull request.

zPereiraa committed:
feat(domain): implemented Edit Enrollment

related issues: #28

zPereiraa committed:
feat(service): implemented Edit Enrollment

zPereiraa committed:
feat(controller): implemented PUT

related issues: #32

AndreGamito1 committed:
feat: added update on participation.java

AndreGamito1 committed:
feat: added Update Participation Method Test

Related: #52

AndreGamito1 committed:
feat: added updateParticipation in ParticipationService.java

Related: #48

AndreGamito1 committed:
feat: added UpdateParticipation Service test

Related: #50

AndreGamito1 committed:
feat: added updateParticipation method in ParticipationController.java

Related: #44

AndreGamito1 committed:
feat: added UpdateParticipationWebServiceIT.groovy

Related: #46

AndreGamito1 committed:
feat: added deleteParticipation in ParticipationService.java

Related: #49

AndreGamito1 committed:
feat: added deleteParticipation method in ParticipationController.java

Related: #45

AndreGamito1 committed:
feat: added deleteParticipation Service test

Related: #51

AndreGamito1 committed:
feat: added deleteParticipation WebService Test

Related: #47

zPereiraa committed:
feat(domain test): implemented UpdateEnrollmentMethodTest

zPereiraa committed:
feat(Service Test): implemented UpdateEnrollmentServiceTest

related issues: #40

zPereiraa committed:
feat(domain): implemented Delete Enrollment and fix Edit Enrollment

related issues: #28 #29

zPereiraa committed:
feat(WebService Test): implemented UpdateEnrollmentWebServiceIT

related issues: #38

AndreGamito1 committed:
feat: added delete method in ParticipationService.java and updated necessary files to support delete operation

Related: #43

AndreGamito1 committed:
feat: added delete participation domain tests

Related: #53

zPereiraa committed:
feat(controller): implemented DELETE

related issues: #33

zPereiraa committed:
feat(domain): fix Delete and Edit

related issues: #28 #29

zPereiraa committed:
feat(Service): implemented Delete Enrollment

related issues: #35

Commit:
feat(domain test): updated UpdateEnrollmentMethodTest

related issues: #36

zPereiraa committed:
feat(domain/serice/webservice tests):

	-implemented DeleteEnrollmentMethodTest
	-implemented DeleteEnrollmentServiceTest
	-updated UpdateEnrollmentServiceTest
	-implemented DeleteEnrollmentWebService

related issues: #37 #39 #40 #41

zPereiraa committed:
feat(): small changes in some files

zPereiraa committed:
feat(fix): small change

zPereiraa committed:
feat(fix): small changes

zPereiraa committed:
feat(fix): small changes

zPereiraa committed:
feat(fix): fix some errors

zPereiraa committed:
feat(fix): fix some erros

zPereiraa committed:
feat(domain/service tests): fix some errors

relating issues: #37 #36 #35 #41

zPereiraa committed:
feat(fix): small changes

AndreGamito1 committed:
fix: prevent unauthorized access in participation controller

ritosilva committed:
fix update and remove participation access control

AndreGamito1 committed:
fix: fixed webservice tests

ritosilva committed:
Merge branch 'gamito' into pereira

ritosilva commented on backend/src/main/java/pt/ulisboa/tecnico/socialsoftware/humanaethica/enrollment/domain/Enrollment.java:
```diff
@@ -34,6 +34,22 @@ public Enrollment(Activity activity, Volunteer volunteer, EnrollmentDto enrollme
     }
 
+    public void edit(EnrollmentDto enrollmentDto) {  
+        editOrDeleteEnrollmentBeforeDeadline();
```
This should be in an invariant that is verified at the end

ritosilva commented on backend/src/main/java/pt/ulisboa/tecnico/socialsoftware/humanaethica/enrollment/domain/Enrollment.java:
```diff
@@ -34,6 +34,22 @@ public Enrollment(Activity activity, Volunteer volunteer, EnrollmentDto enrollme
+        setMotivation(enrollmentDto.getMotivation());
+
+        motivationIsRequired();
+
```
verify invariants at the end

ritosilva commented on backend/src/main/java/pt/ulisboa/tecnico/socialsoftware/humanaethica/enrollment/domain/Enrollment.java:
```diff
@@ -34,6 +34,22 @@ public Enrollment(Activity activity, Volunteer volunteer, EnrollmentDto enrollme
+    public void edit(EnrollmentDto enrollmentDto) {  
+        editOrDeleteEnrollmentBeforeDeadline();
+
+        setEnrollmentDateTime(LocalDateTime.now());
```
Keep the time of the creation of the enrollment

ritosilva commented on backend/src/main/java/pt/ulisboa/tecnico/socialsoftware/humanaethica/enrollment/domain/Enrollment.java:
```diff
@@ -34,6 +34,22 @@ public Enrollment(Activity activity, Volunteer volunteer, EnrollmentDto enrollme
+    }
+
+    public void delete(){
+        editOrDeleteEnrollmentBeforeDeadline();
```
verify invariants at the end

ritosilva commented on backend/src/main/java/pt/ulisboa/tecnico/socialsoftware/humanaethica/enrollment/EnrollmentController.java:
```diff
@@ -37,4 +37,18 @@ public EnrollmentDto createEnrollment(Principal principal, @PathVariable Integer
     }
+
+    @PutMapping("/activities/{activityId}/enrollments/{enrollmentId}")
+    @PreAuthorize("(hasRole('ROLE_VOLUNTEER'))")
```
It is necessary to verify that it is the volunteer of this enrollment

ritosilva commented on backend/src/main/java/pt/ulisboa/tecnico/socialsoftware/humanaethica/enrollment/EnrollmentController.java:
```diff
@@ -37,4 +37,18 @@ public EnrollmentDto createEnrollment(Principal principal, @PathVariable Integer
+    }
+
+    @DeleteMapping("/activities/{activityId}/enrollments/{enrollmentId}")
+    @PreAuthorize("(hasRole('ROLE_VOLUNTEER'))")
```
It is necessary to verify that it is the volunteer of this enrolment

ritosilva commented on backend/src/main/java/pt/ulisboa/tecnico/socialsoftware/humanaethica/enrollment/EnrollmentController.java:
```diff
@@ -37,4 +37,18 @@ public EnrollmentDto createEnrollment(Principal principal, @PathVariable Integer
         return enrollmentService.createEnrollment(userId, activityId, enrollmentDto);
     }
+
+    @PutMapping("/activities/{activityId}/enrollments/{enrollmentId}")
```
activityId is not necessary, use only "/enrollments/{enrollmentId}"

ritosilva commented on backend/src/main/java/pt/ulisboa/tecnico/socialsoftware/humanaethica/enrollment/EnrollmentController.java:
```diff
@@ -37,4 +37,18 @@ public EnrollmentDto createEnrollment(Principal principal, @PathVariable Integer
+        return enrollmentService.editEnrollment(enrollmentId, enrollmentDto);
+    }
+
+    @DeleteMapping("/activities/{activityId}/enrollments/{enrollmentId}")
```
activityId is not necessary, use only "/enrollments/{enrollmentId}"

ritosilva commented on backend/src/main/java/pt/ulisboa/tecnico/socialsoftware/humanaethica/enrollment/EnrollmentService.java:
```diff
@@ -62,4 +62,32 @@ public EnrollmentDto createEnrollment(Integer userId, Integer activityId, Enroll
     }
+
+    @Transactional(isolation = Isolation.READ_COMMITTED)
+    public EnrollmentDto editEnrollment(Integer enrollmentId, EnrollmentDto enrollmentDto) {
```
use the name update instead of edit, to be consistent with other code

ritosilva commented on backend/src/main/java/pt/ulisboa/tecnico/socialsoftware/humanaethica/enrollment/EnrollmentService.java:
```diff
@@ -62,4 +62,32 @@ public EnrollmentDto createEnrollment(Integer userId, Integer activityId, Enroll
+
+        Enrollment enrollment = enrollmentRepository.findById(enrollmentId).orElseThrow(() -> new HEException(ENROLLMENT_NOT_FOUND, enrollmentId));
+
+        enrollment.edit(enrollmentDto);
```
use update

ritosilva reviewed (changes requested):
Do the requested changes

zPereiraa committed:
feat(domain/service/controller/tests):
      - changed EnrollmentController.java
      - implemented "ENROLLMENT.MANAGER" in HEPermissionEvaluator.java
      - changed Enrollment.java
      -  some changes in UpdateEnrollmentMethodTest, UpdateEnrollmentServiceTest, DeleteEnrollmentWebServiceIT and UpdateEnrollmentWebServiceIT

realted issues: #32 #33 #28 #29 #40 #36

ritosilva commented on backend/src/main/java/pt/ulisboa/tecnico/socialsoftware/humanaethica/enrollment/domain/Enrollment.java:
```diff
@@ -34,6 +34,20 @@ public Enrollment(Activity activity, Volunteer volunteer, EnrollmentDto enrollme
+        setMotivation(enrollmentDto.getMotivation());
+
+        motivationIsRequired();
+        editOrDeleteEnrollmentBeforeDeadline();
```
here verify invariants to check that the update preserves de invariants

ritosilva commented on backend/src/main/java/pt/ulisboa/tecnico/socialsoftware/humanaethica/enrollment/domain/Enrollment.java:
```diff
@@ -34,6 +34,20 @@ public Enrollment(Activity activity, Volunteer volunteer, EnrollmentDto enrollme
+        volunteer.removeEnrollment(this);
+        activity.removeEnrollment(this);
+
+        editOrDeleteEnrollmentBeforeDeadline();
```
here verifyInvariants to check that the delete preserves the invariant

zPereiraa committed:
feat(domain): small changes

ritosilva committed:
Merge branch 'master' into pereira
